Make the subscriber in awaitOne less permissive

dkhalanskyjb opened the pull request:
Reactive Streams' Subscriber's implementation for `await*` operations was assuming that the publisher is correct. Now, the implementation detects some instances of problematic behavior for publishers and reports them.

Fixes https://github.com/Kotlin/kotlinx.coroutines/issues/2079

dkhalanskyjb commented on reactive/kotlinx-coroutines-reactive/test/IntegrationTest.kt:
```diff
@@ -130,6 +132,73 @@ class IntegrationTest(
+            assertFailsWith<IllegalStateException> { publisher(block).operation() }
+
+        // Rule 1.1 broken: the publisher produces more values than requested.
+        assertDetectsBadPublisher<Int>({ awaitFirst() }) {
```
Is it expected that `::awaitFirst()` doesn't work here?

qwwdfsad commented on reactive/kotlinx-coroutines-reactive/src/Await.kt:
```diff
@@ -134,29 +135,47 @@ private suspend fun <T> Publisher<T>.awaitOne(
             subscription = sub
             cont.invokeOnCancellation { sub.cancel() }
-            sub.request(if (mode == Mode.FIRST) 1 else Long.MAX_VALUE)
+            sub.request(if (mode == Mode.FIRST || mode == Mode.FIRST_OR_DEFAULT) 1 else Long.MAX_VALUE)
```
Shouldn't it be anything but `Mode.LAST`?

dkhalanskyjb replied:
I don't think so, no: correct publishers don't produce more values than requested, and if we only requested one value for `SINGLE`, there wouldn't be any difference between `FIRST` and `SINGLE`.

qwwdfsad commented on reactive/kotlinx-coroutines-reactive/src/Await.kt:
```diff
@@ -134,29 +135,47 @@ private suspend fun <T> Publisher<T>.awaitOne(
 
         override fun onNext(t: T) {
+            val sub = subscription.checkInitialized("onNext")
+            subscription.checkInitialized("onNext")
```
Double `checkInitialized` seems to be redundant

qwwdfsad commented on reactive/kotlinx-coroutines-reactive/src/Await.kt:
```diff
@@ -134,29 +135,47 @@ private suspend fun <T> Publisher<T>.awaitOne(
-                    }
+                    if (seenValue)
+                        // TODO: decide if we want to be lenient here: after all, nothing breaks if this isn't true
+                        moreThanOneValueProvidedException(mode)
```
When writing a comment along with a statement, please wrap it in curly braces, otherwise it's a bit hard to read

qwwdfsad replied:
Throwing here breaks `#2.13`. This is expected, but still worth properly documenting to know why it is here even tho it violates the spec.
Also, removing this condition does not trigger any test failures, because double-resume also throws `IllegalStateException`. It's worth verifying the exception message at least once

dkhalanskyjb replied:
I am now verifying the exception messages. I don't like what I did in order to achieve it, but I like copy-pasting exception messages into tests even less, given how brittle that approach is. Maybe you have some better suggestions?

qwwdfsad replied:
The approach we're consistently using is the following:

- Do not expose exception messages to test in any way (e.g. via `internal` function)
- Instead of strictly comparing the error message as a whole, typically it's enough to do something like `e.message!!.contains("some_key_word_or_meaningful_part_of_the_message")`. 
 
 It simplifies the maintenance a bit, reduces the number of functions/string consts, but it's still brittle. Yet exception messages are rarely changed and the purpose of the test is rather fix the current behaviour

qwwdfsad commented on reactive/kotlinx-coroutines-reactive/src/Await.kt:
```diff
@@ -169,6 +188,8 @@ private suspend fun <T> Publisher<T>.awaitOne(
 
         @Suppress("UNCHECKED_CAST")
         override fun onComplete() {
+            subscription.checkInitialized("onComplete") // TODO: maybe don't enforce?
```
It seems like other implementations from reactive libraries do not enforce such rule, maybe we could stick with that as well

qwwdfsad commented on reactive/kotlinx-coroutines-reactive/test/IntegrationTest.kt:
```diff
@@ -148,35 +148,41 @@ class IntegrationTest(
+            block: Subscriber<in T>.(n: Long) -> Unit,
+        ) =
+            assertFailsWith<IllegalStateException> { publisher(block).operation() }.let {
+                message == null || it.message == message
```
This line doesn't actually test exception messages, the resulting boolean is ignored

dkhalanskyjb replied:
Yeah, sorry. Fixed in the commit that changes the whole error-passing mechanism here.

dkhalanskyjb commented:
@qwwdfsad could you please check the latest commit? I spotted several errors during the final review, fixed them, and added a test to verify a new statement in a comment.
